File: code/python/algorithm_thinking_article-01.py
"""BFS Algorithm Problem Solving Routine Framework"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().openLock(*args, **kwargs)


class _Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        deads = set(deadends)
        visited = set()
        q = collections.deque()
        step = 0
        q.append("0000")
        visited.add("0000")
        while q:
            sz = len(q)
            for i in range(sz):
                cur = q.popleft()
                if cur in deads:
                    continue
                if cur == target:
                    return step
                for j in range(4):
                    up = plusOne(cur,j)
                    if up not in visited:
                        q.append(up)
                        visited.add(up)
                    down = minusOne(cur,j)
                    if down not in visited:
                        q.append(down)
                        visited.add(down)
            step += 1
        return -1
    def plusOne(s: str, j: int) -> str:
        ch = list(s)
        if ch[j] == '9':
            ch[j] = '0'
        else:
            ch[j] = chr(ord(ch[j])+1)
        return "".join(ch)
    def minusOne(s: str, j: int) -> str:
        ch = list(s)
        if ch[j] == '0':
            ch[j] = '9'
        else:
            ch[j] = chr(ord(ch[j])-1)
        return "".join(ch)

File: code/javascript/algorithm_thinking_article-01.js
/**
 * BFS Algorithm Problem Solving Routine Framework
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return plusOne(...args);
}

var openLock = function(deadends, target) {
  const deads = new Set(deadends);
  const visited = new Set();
  const q = [];
  let step = 0;
  q.push("0000");
  visited.add("0000");
  while (q.length > 0) {
    let sz = q.length;
    for (let i = 0; i < sz; i++) {
      let cur = q.shift();
      if (deads.has(cur)) continue;
          if (cur === target) return step;
      for (let j = 0; j < 4; j++) {
        let up = plusOne(cur, j);
        if (!visited.has(up)) {
          q.push(up);
          visited.add(up);
        }
        let down = minusOne(cur, j);
        if (!visited.has(down)) {
          q.push(down);
          visited.add(down);
        }
      }
    }
    step++;
  }
  return -1;
}
function plusOne(s, j) {
  let ch = s.split("");
  if (ch[j] === "9") ch[j] = "0";
  else ch[j] = String(+ch[j]+1);
  return ch.join("");
}
function minusOne(s, j) {
  let ch = s.split("");
  if (ch[j] === "0") ch[j] = "9";
  else ch[j] = String(+ch[j]-1);
  return ch.join("");
}
